File: DialedInCoffee/RulesEngine/LearningHeuristics.swift
import Foundation

public enum LearningHeuristics {
    /// True when there are 3+ recent adjustments to the same parameter in the same direction
    /// with no "good" outcome between them — the user is chasing their tail.
    public static func detectRepeatedFailure(history: [HistorySnapshot]) -> Bool {
        guard history.count >= 3 else { return false }
        let recent = Array(history.suffix(5))

        // Group by parameter; bail if any "good" outcome appears recently.
        if recent.contains(where: { $0.outcome == .good }) { return false }

        // Find the most-touched parameter in recent history.
        let counts = Dictionary(grouping: recent, by: { $0.primaryAdjustmentParameter })
            .mapValues { $0.count }
        guard let (param, count) = counts.max(by: { $0.value < $1.value }), count >= 3 else { return false }

        // Check those touches all move in the same direction.
        let touches = recent.filter { $0.primaryAdjustmentParameter == param }
        return monotonicallySameDirection(touches, parameter: param)
    }

    /// Most recent "good"-marked snapshot, if any.
    public static func lastGood(history: [HistorySnapshot]) -> HistorySnapshot? {
        history.reversed().first { $0.outcome == .good }
    }

    private static func monotonicallySameDirection(
        _ snapshots: [HistorySnapshot],
        parameter: AdjustmentParameter
    ) -> Bool {
        var lastDelta: Int? = nil
        for snap in snapshots {
            let delta: Int
            switch parameter {
            case .grinder: delta = snap.afterSettings.grinder - snap.beforeSettings.grinder
            case .strength: delta = snap.afterSettings.strength - snap.beforeSettings.strength
            case .volume: delta = snap.afterSettings.volumeML - snap.beforeSettings.volumeML
            case .milkDuration: delta = snap.afterSettings.milkSeconds - snap.beforeSettings.milkSeconds
            case .temperature:
                let order: [TemperatureLevel] = [.low, .normal, .high]
                let bi = order.firstIndex(of: snap.beforeSettings.temperature) ?? 1
                let ai = order.firstIndex(of: snap.afterSettings.temperature) ?? 1
                delta = ai - bi
            case .beans:
                return false
            }
            if delta == 0 { return false }
            if let last = lastDelta {
                if (last > 0) != (delta > 0) { return false }
            }
            lastDelta = delta
        }
        return lastDelta != nil
    }
}

File: android/core/src/main/kotlin/com/cawfee/domain/rules/LearningHeuristics.kt
package com.cawfee.domain.rules

import com.cawfee.domain.model.AdjustmentOutcome
import com.cawfee.domain.model.AdjustmentParameter
import com.cawfee.domain.model.HistorySnapshot

/** Detects "chasing your tail" patterns. Ported from LearningHeuristics.swift. */
object LearningHeuristics {

    /**
     * True when there are 3+ recent adjustments to the same parameter in the same
     * direction with no "good" outcome between them.
     */
    fun detectRepeatedFailure(history: List<HistorySnapshot>): Boolean {
        if (history.size < 3) return false
        val recent = history.takeLast(5)

        if (recent.any { it.outcome == AdjustmentOutcome.GOOD }) return false

        val counts = recent.groupingBy { it.primaryAdjustmentParameter }.eachCount()
        val top = counts.maxByOrNull { it.value } ?: return false
        if (top.value < 3) return false

        val touches = recent.filter { it.primaryAdjustmentParameter == top.key }
        return monotonicallySameDirection(touches, top.key)
    }

    /** Most recent "good"-marked snapshot in array order, if any. */
    fun lastGood(history: List<HistorySnapshot>): HistorySnapshot? =
        history.asReversed().firstOrNull { it.outcome == AdjustmentOutcome.GOOD }

    private fun monotonicallySameDirection(
        snapshots: List<HistorySnapshot>,
        parameter: AdjustmentParameter,
    ): Boolean {
        var lastDelta: Int? = null
        for (snap in snapshots) {
            val delta = when (parameter) {
                AdjustmentParameter.GRINDER -> snap.afterSettings.grinder - snap.beforeSettings.grinder
                AdjustmentParameter.STRENGTH -> snap.afterSettings.strength - snap.beforeSettings.strength
                AdjustmentParameter.VOLUME -> snap.afterSettings.volumeML - snap.beforeSettings.volumeML
                AdjustmentParameter.MILK_DURATION -> snap.afterSettings.milkSeconds - snap.beforeSettings.milkSeconds
                AdjustmentParameter.TEMPERATURE ->
                    snap.afterSettings.temperature.order - snap.beforeSettings.temperature.order
                AdjustmentParameter.BEANS -> return false
            }
            if (delta == 0) return false
            val last = lastDelta
            if (last != null && (last > 0) != (delta > 0)) return false
            lastDelta = delta
        }
        return lastDelta != null
    }
}
